tests/multi_ptr/multi_ptr_subscript_op_common.h

```diff
@@ -0,0 +1,166 @@
+
+          // standard indexation
+          for (int i = 0; i < array_size; i++) {
+            ret_arr_acc[i] = multi_ptr[i];
+          }
+
+          // negative indexation
+          for (int i = 0; i < array_size; i++) {
+            ret_arr_neg_acc[i] = multi_ptr_negative[-i];
+          }
+
+          // type check
+          ret_type_acc[0] = std::is_same_v<decltype(multi_ptr[0]),
+                                           typename multi_ptr_t::reference>;
+        };
+
+        if constexpr (space == sycl::access::address_space::local_space) {
+          sycl::local_accessor<T> exp_arr_acc{{array_size}, cgh};
+          sycl::local_accessor<T> exp_arr_neg_acc{{array_size}, cgh};
+          cgh.single_task([=] {
+            value_operations::assign(exp_arr_acc, exp_arr);
+            value_operations::assign(exp_arr_neg_acc, exp_arr);
+
+            // pointer to the end of the array
+            T *arr_end = &exp_arr_neg_acc[array_size - 1];
+            multi_ptr_t multi_ptr_negative =
+                sycl::address_space_cast<space, decorated>(arr_end);
+
+            test_device_code(exp_arr_acc, multi_ptr_negative);
+          });
+        } else if constexpr (space ==
+                             sycl::access::address_space::private_space) {
+          cgh.single_task([=] {
+            T *priv_arr = const_cast<T *>(&(exp_arr[0]));
+            multi_ptr_t priv_arr_mptr =
+                sycl::address_space_cast<space, decorated>(priv_arr);
+            multi_ptr_t multi_ptr_negative =
+                sycl::address_space_cast<space, decorated>(priv_arr +
+                                                           array_size - 1);
+            test_device_code(priv_arr_mptr, multi_ptr_negative);
+          });
+        } else {
+          auto exp_arr_acc =
+              exp_arr_buffer.template get_access<sycl::access_mode::read>(cgh);
+          auto exp_arr_neg_acc =
+              exp_arr_buffer.template get_access<sycl::access_mode::read>(cgh);
+          cgh.single_task([=] {
+            T *arr_end = const_cast<T *>(&exp_arr_neg_acc[array_size - 1]);
+            multi_ptr_t multi_ptr_negative =
+                sycl::address_space_cast<space, decorated>(arr_end);
+            test_device_code(exp_arr_acc, multi_ptr_negative);
+          });
+        }
+      });
+    }
+
+    std::reverse(ret_arr_neg.begin(), ret_arr_neg.end());
+
+    // multi_ptr::operator[](std::ptrdiff_t) available only when:
+    // !std::is_void<sycl::multi_ptr::value_type>::value
+    if constexpr (!std::is_void_v<typename multi_ptr_t::value_type>) {
+      INFO("Check operator[](std::ptrdiff_t) return value and type"
+           << "T" << type_name << "address_space" << address_space_name
+           << "decorated" << is_decorated_name);
+      CHECK(ret_arr == exp_arr);
+      CHECK(ret_arr_neg == exp_arr);
+      CHECK(ret_type_check);
+    }
+  }
+};
+
+template <typename T>
+class check_multi_ptr_subscript_op {
```

Comment:
It is unclear to me why this is not just a function instead of class + `operator()`.

Reply:
OK thanks for the clarification.